File: public/code/indicators__chandelier.py
"""
Chandelier Exit Strategy
========================
ATR tabanlı trailing stop sistemi.

Long stop = highest(high, N) - ATR(N) * multiplier
Short stop = lowest(low, N) + ATR(N) * multiplier

Long entry: fiyat short stop'un üstüne kırılırsa.
Short entry: fiyat long stop'un altına kırılırsa.
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, atr, ema


def detect_chandelier_signal(
    df: pd.DataFrame,
    period: int = 22,
    multiplier: float = 3.0,
    use_trend_filter: bool = True,
    ema_period: int = 200,
) -> dict:
    if len(df) < max(period, ema_period) + 5:
        return make_signal(reason="Insufficient data")

    atr_v = atr(df, period)
    long_stop = df["high"].rolling(period).max() - atr_v * multiplier
    short_stop = df["low"].rolling(period).min() + atr_v * multiplier
    trend_ema = ema(df["close"], ema_period)

    current = df.iloc[-1]
    prev = df.iloc[-2]
    current_atr = atr_v.iloc[-1]

    # Long entry: fiyat short_stop'un üstüne çıktı
    long_breakout = current["close"] > short_stop.iloc[-1] and prev["close"] <= short_stop.iloc[-2]
    short_breakout = current["close"] < long_stop.iloc[-1] and prev["close"] >= long_stop.iloc[-2]

    if long_breakout:
        if use_trend_filter and current["close"] < trend_ema.iloc[-1]:
            return make_signal(reason="Chandelier long but downtrend")
        entry = current["close"]
        stop_loss = long_stop.iloc[-1]
        risk = entry - stop_loss
        if risk <= 0:
            return make_signal(reason="Invalid risk")
        return make_signal(
            signal="long",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry + risk * 1.5, entry + risk * 2.5, entry + risk * 4],
            confidence=0.7,
            reason="Chandelier long breakout",
            metadata={"long_stop": float(stop_loss)},
        )

    if short_breakout:
        if use_trend_filter and current["close"] > trend_ema.iloc[-1]:
            return make_signal(reason="Chandelier short but uptrend")
        entry = current["close"]
        stop_loss = short_stop.iloc[-1]
        risk = stop_loss - entry
        if risk <= 0:
            return make_signal(reason="Invalid risk")
        return make_signal(
            signal="short",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry - risk * 1.5, entry - risk * 2.5, entry - risk * 4],
            confidence=0.7,
            reason="Chandelier short breakout",
            metadata={"short_stop": float(stop_loss)},
        )

    return make_signal(reason="No chandelier breakout")

File: public/code/indicators__chandelier.ts
/**
 * Chandelier Exit Strategy - TypeScript
 */
import { Candle, Signal, makeSignal, atr, ema } from "../../common";

export function detectChandelierSignal(
  candles: Candle[],
  opts: { period?: number; multiplier?: number; useTrendFilter?: boolean; emaPeriod?: number } = {}
): Signal {
  const { period = 22, multiplier = 3, useTrendFilter = true, emaPeriod = 200 } = opts;
  if (candles.length < Math.max(period, emaPeriod) + 5)
    return makeSignal({ reason: "Insufficient data" });

  const atrV = atr(candles, period);
  const highs = candles.map((c) => c.high);
  const lows = candles.map((c) => c.low);
  const closes = candles.map((c) => c.close);

  const longStop: number[] = [];
  const shortStop: number[] = [];
  for (let i = 0; i < candles.length; i++) {
    if (i < period - 1) { longStop.push(NaN); shortStop.push(NaN); continue; }
    const maxH = Math.max(...highs.slice(i - period + 1, i + 1));
    const minL = Math.min(...lows.slice(i - period + 1, i + 1));
    longStop.push(maxH - atrV[i] * multiplier);
    shortStop.push(minL + atrV[i] * multiplier);
  }

  const emaV = ema(closes, emaPeriod);
  const i = candles.length - 1;
  const current = candles[i];
  const prev = candles[i - 1];

  const longBreakout = current.close > shortStop[i] && prev.close <= shortStop[i - 1];
  const shortBreakout = current.close < longStop[i] && prev.close >= longStop[i - 1];

  if (longBreakout) {
    if (useTrendFilter && current.close < emaV[i]) return makeSignal({ reason: "Long but downtrend" });
    const entry = current.close;
    const stopLoss = longStop[i];
    const risk = entry - stopLoss;
    if (risk <= 0) return makeSignal({ reason: "Invalid risk" });
    return makeSignal({
      signal: "long", entry, stop_loss: stopLoss,
      take_profit: [entry + risk * 1.5, entry + risk * 2.5, entry + risk * 4],
      confidence: 0.7, reason: "Chandelier long breakout",
      timestamp: current.timestamp, metadata: { long_stop: stopLoss },
    });
  }
  if (shortBreakout) {
    if (useTrendFilter && current.close > emaV[i]) return makeSignal({ reason: "Short but uptrend" });
    const entry = current.close;
    const stopLoss = shortStop[i];
    const risk = stopLoss - entry;
    if (risk <= 0) return makeSignal({ reason: "Invalid risk" });
    return makeSignal({
      signal: "short", entry, stop_loss: stopLoss,
      take_profit: [entry - risk * 1.5, entry - risk * 2.5, entry - risk * 4],
      confidence: 0.7, reason: "Chandelier short breakout",
      timestamp: current.timestamp, metadata: { short_stop: stopLoss },
    });
  }
  return makeSignal({ reason: "No chandelier breakout" });
}
